### PR TITLE
SWDEV-220503:  this_grid().thread_rank() gives incorrect result

### PR DESCRIPTION
**Pull request by mshivama**
* There was a bug while computing this_grid().thread_rank(), fixed it.
* The semantics of this_grid().thread_rank() computation is as follows.
   1.  Compute number of work-groups to be passed before arriving at current work-group
   2.  Using above, compute total number of threads being passed till current work-group
   3.  Computer local thread-rank (of current thread) within current work-group
   4.  Return sum of above 2 and 3.

### CODE REVIEW

**Comment by mshivama** [comment]
Hello @b-sumner @asalmanp and @jlgreathouse,

Could you please review this PR which fixes issue raised in SWDEV-220503.

File: include/hip/hcc_detail/hip_cooperative_groups_helper.h [hunk 1]
```diff
@@ -106,21 +106,21 @@ __CG_STATIC_QUALIFIER__ uint32_t size() {
 }
 
 __CG_STATIC_QUALIFIER__ uint32_t thread_rank() {
-  // Compute global id of the workgroup to which the current threads belongs to
+  // Compute global id of the workgroup to which the current thread belongs to
   uint32_t blkIdx =
            (uint32_t)((hipBlockIdx_z * hipGridDim_y * hipGridDim_x) +
-                      (hipBlockIdx_y * hipGridDim_x) +
+                      (hipBlockIdx_x * hipGridDim_y) +
```
Comment (jlgreathouse) [inline include/hip/hcc_detail/hip_cooperative_groups_helper.h]: This does not seem right. I believe the previous version was correct. Could you describe why this change would be correct?

And *if* this change is correct (which I don't think it is), then why not change the line above as well? Why remove hipBlockIdx_y from the calculation, but leave hipBlockIdx_z?

File: include/hip/hcc_detail/hip_cooperative_groups_helper.h [hunk 2]
```diff
@@ -106,21 +106,21 @@ __CG_STATIC_QUALIFIER__ uint32_t size() {
...
   // within grid
   uint32_t num_threads_till_current_workgroup =
-           (uint32_t)(blkIdx * (hipBlockIdx_x * hipBlockIdx_y * hipBlockIdx_z));
+           (uint32_t)(blkIdx * (hipBlockDim_x * hipBlockDim_y * hipBlockDim_z));
 
   // Compute thread local rank within current workgroup
   uint32_t local_thread_rank =
            (uint32_t)((hipThreadIdx_z * hipBlockDim_y * hipBlockDim_x) +
-                      (hipThreadIdx_y * hipBlockDim_x) +
+                      (hipThreadIdx_x * hipBlockDim_y) +
```
Comment (jlgreathouse) [inline include/hip/hcc_detail/hip_cooperative_groups_helper.h]: See my comment above about the similar change on line 122. I disagree with this change.

File: include/hip/hcc_detail/hip_cooperative_groups_helper.h (hunk 1 above)
Comment (mshivama) [reply include/hip/hcc_detail/hip_cooperative_groups_helper.h]: Actually I got confused :) Let me think again, and correct it in case, if I made mistake.

File: include/hip/hcc_detail/hip_cooperative_groups_helper.h [hunk 3]
```diff
@@ -106,21 +106,21 @@ __CG_STATIC_QUALIFIER__ uint32_t size() {
...
            (uint32_t)((hipBlockIdx_z * hipGridDim_y * hipGridDim_x) +
-                      (hipBlockIdx_y * hipGridDim_x) +
+                      (hipBlockIdx_x * hipGridDim_y) +
                       (hipBlockIdx_x));
 
   // Compute total number of threads being passed to reach current workgroup
   // within grid
   uint32_t num_threads_till_current_workgroup =
-           (uint32_t)(blkIdx * (hipBlockIdx_x * hipBlockIdx_y * hipBlockIdx_z));
+           (uint32_t)(blkIdx * (hipBlockDim_x * hipBlockDim_y * hipBlockDim_z));
```
Comment (jlgreathouse) [inline include/hip/hcc_detail/hip_cooperative_groups_helper.h]: I believe this change alone is sufficient. I tested with a 3D kernel launch and this change (and *only* this change, not the other two changes) is sufficient for correct execution.

With the two other changes as well, we get wrong results.

Comment (mshivama) [reply include/hip/hcc_detail/hip_cooperative_groups_helper.h]: Ok I will update it.

**Comment by jlgreathouse** [comment]
LGTM. Before I review-approve this, a question: does the HIP project prefer to do a rebase/squash commit instead of having intermediary commits in the repo? On most large projects I work with, we don't like having broken commits in the middle of the stream, as it makes bisecting difficult.

**Comment by mshivama** [comment]
Hello @mangupta, Could you please merge this PR to HIP master?

**Comment by mangupta** [comment]
> LGTM. Before I review-approve this, a question: does the HIP project prefer to do a rebase/squash commit instead of having intermediary commits in the repo? On most large projects I work with, we don't like having broken commits in the middle of the stream, as it makes bisecting difficult.

@jlgreathouse Yes. We prefer to squash-commit the PRs to avoid broken commits in the history.
